### DISCUSSION TITLE
Question about Step 21. C code output to UART

### DISCUSSION THREAD

**Discussion by pongsagon** [BrunoLevy/learn-fpga#122] [General]
In assembly, we can set the memory address to point to UART to output to it.
When we write in C using putchar, how does it know our custom defined UART memory mapped address?
Thank you very much
Matt

**Comment by yrabbit** [comment]
Because they are clearly stated in the putchar sources?

https://github.com/BrunoLevy/learn-fpga/blob/master/FemtoRV/TUTORIALS/FROM_BLINKER_TO_RISCV/FIRMWARE/putchar.S

**Reply by pongsagon** [reply]
Yes, the putchar asm code seem to be there as you have mentioned.
I am not quite understand about the makefile and compile process.
Does it mean that when we compile C, we set the compiler flag so that it uses the custom putchar.s and not the stdlib?

Thanks

**Reply by yrabbit** [reply]
When linking, functions from explicitly specified object files take precedence over library functions. The object files listed in this variable will take precedence.

https://github.com/BrunoLevy/learn-fpga/blob/5bb5c5e5490d31366e41e16d8b5822a62d70254a/FemtoRV/TUTORIALS/FROM_BLINKER_TO_RISCV/FIRMWARE/Makefile#L7

**Reply by pongsagon** [reply]
Got it.
Thank you very much for your reply.

**Comment by BrunoLevy** [comment]
In fact, the standard library is not linked at all, thanks to the `-nostdlib` flag, so the only version of `putchar()` and `printf()` are the ones that are explicitly linked (I would not trust too much the rules of precedence for that).

**Reply by pongsagon** [reply]
I see.
Thanks a lot for this great tutorial.
